dbschema/common.esdl

```diff
@@ -27,8 +27,4 @@ module default {
   # This will be reflected to dynamically inject portable shapes in our EdgeQL queries.
   function hydrate(typeName: str, scopedValue: json) -> str
     using (typeName);
-  
-  # Get the inclusive upper bound of the given date range.
-  function date_range_get_upper(period: range<cal::local_date>) -> cal::local_date
-    using (<cal::local_date><str>assert_exists(range_get_upper(period)) - <cal::date_duration>"1 day");
```

Comment:
This is still needed. The typecasting workaround was fixed though
```edgeql
  function date_range_get_upper(period: range<cal::local_date>) -> cal::local_date
    using (assert_exists(range_get_upper(period)) - <cal::date_duration>"1 day");
```

Ranges are half closed `[start, end)` meaning the endpoint is not considered to be included.
This matters when we start doing math, like does x date fall within this range?
For example this work week. We see it as
```
[2024-09-02, 2024-09-06]
```
but SQL/EdgeQL stores that as
```
[2024-09-02, 2024-09-06)
```
meaning the 6th is NOT included in that range.

That's why we add one day
```
[2024-09-02, 2024-09-07) == [2024-09-02, 2024-09-06]
```

And that's why this function exists, because when we display the "end date" we want it to be the inclusive endpoint, not the half closed one.

Reply:
I _think_ I understand what you're saying conceptually, but something still isn't adding up for me in practice.

When I run the `create_periodic_report_ranges` in the context of the trigger I created I receive the expected results.  For example, if I call `create_periodic_report_ranges` alone in the editor like this:  

```
create_periodic_report_ranges('2023-01-01', '2023-12-30', 3)
```

I receive this output:

```
[
  {"lower": "2023-01-01T00:00:00", "upper": "2023-03-31T00:00:00", "inc_lower": true, "inc_upper": false},
  {"lower": "2023-04-01T00:00:00", "upper": "2023-06-30T00:00:00", "inc_lower": true, "inc_upper": false},
  {"lower": "2023-07-01T00:00:00", "upper": "2023-09-30T00:00:00", "inc_lower": true, "inc_upper": false},
  {"lower": "2023-10-01T00:00:00", "upper": "2023-12-31T00:00:00", "inc_lower": true, "inc_upper": false},
  ...
]
```

which seems accurate and what we'd want.

Now if I actually run the trigger and then look at the data on the `start` and `end` fields on the `Financial Report` in the data explorer I see this:

```
[
  {"lower": "2023-01-01T00:00:00", "upper": "2023-03-30T00:00:00", "inc_lower": true, "inc_upper": false},
  {"lower": "2023-04-01T00:00:00", "upper": "2023-06-29T00:00:00", "inc_lower": true, "inc_upper": false},
  {"lower": "2023-07-01T00:00:00", "upper": "2023-09-29T00:00:00", "inc_lower": true, "inc_upper": false},
  {"lower": "2023-10-01T00:00:00", "upper": "2023-12-30T00:00:00", "inc_lower": true, "inc_upper": false},
  ...
]
```

The  end dates seem to be one less than what I would expect to see.  I understand that they are calculated values applying the custom `date_range_get_upper` function and thus subtracting a day, but is that what we want?

I do see that the `period` on the `Financial Report` is correct, however.

...

I guess the only thing I can think of is that I'm incorrect in my assessment that we really want a financial reporting period to look like this:  `2023-01-01 -- 2023-03-31`.  Perhaps conceptually we want to think of the _period_ as `2023-01-01 -- 2023-04-01`?  If that's the case then I'm just creating the periods in my function with an end date 1 less than we desire?